### PR TITLE
Add NormalizeScore extension point for scheduler framework.

### LINKED ISSUE #79901: Add Normalize Scoring extension point of the scheduling framework
**What would you like to be added**:
The scheduling framework KEP is now implementable. We have already added plugin interfaces and a few extension points, such as "reserve" and "prebind" plugins.

[Normalize Scoring](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md#normalize-scoring) plugins are used to modify scores before the scheduler computes a final ranking of Nodes. A plugin that registers for this extension point will be called with the scoring results from the same plugin.

This is needed for 1.16.

/sig scheduling
/priority important-soon
ref/ kubernetes/enhancements#624

### LINKED ISSUE #80272: Add unit tests for Scheduler framework
**What would you like to be added**:
Add unit tests for Scheduler [framework](https://github.com/congliu01/kubernetes/blob/73704f1b8e04c573185f230019cc1891375c283b/pkg/scheduler/framework/v1alpha1/framework.go#L264). 

Currently there are only [integration tests](https://github.com/congliu01/kubernetes/blob/c54c4d1962d84b0659644f8ae9c9976b547d021f/test/integration/scheduler/framework_test.go#L435-L434) for the framework. 

**Why is this needed**:
* Unit tests are easier to write and maintain. 

The integration tests requires spawning up etcd and the scheduler reconcile loop, which includes way more logic than the framework itself. It's also hard to control and inspect the input/output easily due to the fact that the Scheduler is a "black box" during integration test. As a result, we can find tests that are really hard to understand and maintain, see [example](https://github.com/kubernetes/kubernetes/issues/80227)

* Unit tests can run fast and deterministically.

Due to the asynchronous and "black box" nature of running the Scheduler reconcile loop, the current integration tests takes longer time and flaky. Here is a recent [example](https://github.com/kubernetes/kubernetes/issues/80255) of this.

* Unit tests can potentially cover more cases than integration tests.

Again, it's hard to control and inspect the system state running integration test. For example, it's not possible to inspect the full output of the Score plugin with integration test.

### PR DESCRIPTION
**Pull request by liu-cong**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Implement normalize plugin extension point so that normalize plugin can be used.

**Which issue(s) this PR fixes**:

Fixes #79901
Also helps with #80272

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Implement normalize plugin extension point for the scheduler framework.

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

[Scheduler Framework](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md)

### CODE REVIEW

**Review by draveness — CHANGES_REQUESTED** [review]
/ok-to-test
/priority important-soon

File: pkg/scheduler/framework/v1alpha1/framework_test.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/framework_test.go]: Do we prefer to use underscores as part of the name in test functions?

**Review by draveness — COMMENTED** [review]
Looks good, but I think partially fixes would close that issue after this gets merged.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (hex108) [inline pkg/scheduler/framework/v1alpha1/framework.go]: We could precheck it when initializing the NormalizeSocrePlugins.

Comment (draveness) [reply pkg/scheduler/framework/v1alpha1/framework.go]: If this is a requirement for running normalizes extension point, we could embed ScorePlugin into NormalizeScorePlugin.

File: pkg/scheduler/framework/v1alpha1/framework_test.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework_test.go]: > Looks good, but I think partially fixes would close that issue after this gets merged.

I think `fixes` is the key word to trigger the close? I can rephrase this to get rid of `fixes`

Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework_test.go]: > Do we prefer to use underscores as part of the name in test functions?

Quickly looked at some other tests and didn't see underscore used. I can remove it.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: I will leave this open for more inputs. I think precheck is good but I am not sure if we should fail the initialization or just print some warning message. 

The current implementation assumes there isn't a necessary match of score plugin list and normalize score plugin list, and it doesn't fail

@salamat, what do you think?

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (bsalamat) [inline pkg/scheduler/framework/v1alpha1/framework.go]: An alternative is to run NormalizeScore plugins in the same `RunScorePlugins`. That would helps save on memory by not passing a large map around, but it would reduce parallelism. I haven't gone through all the issues that it could introduce, so I leave it up to you.

```
```

**Review by bsalamat — COMMENTED** [review]
A few early comments.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: I have discussed this with @ahg-g. As you said, there is a trade-off between saving memory vs. reducing parallelism. I have chosen to optimize for parallelism. Since I am new to scheduler, I don't have enough data points to prove that the tradeoff I made is optimal.

I will keep it as is unless you have further comments.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: This is something I wasn't quite sure about while reading the KEP, thanks for clarifying this!

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: @bsalamat  has clarified this in [this comment](https://github.com/kubernetes/kubernetes/pull/80383#discussion_r306044447). I am going to add precheck logic to make sure that the list of enabled Normalize Score extension points is a subset of Score extension points. And hence the check here is not needed.

Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: I am going to add logic during framework initialization to make sure NormalizeScore extension points are a subset of Score extension points.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (ahg-g) [reply pkg/scheduler/framework/v1alpha1/framework.go]: Weight is currently applied in ```RunScorePlugins```. We need to change that and apply the weight after NormalizeScore, if a NormalizeScore plugin exist.

If we are going with this current code structure, we need to add a ApplyScoreWeights function that gets called after RunNormalizeScorePlugins in generic_scheduler.go.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/framework.go]: We should still test that it actually exist and throw an internal error, it may help us debug bugs (e.g., RunScorePlugins not adding nodeScoreList at all, or Name() return something different since it is still a function).

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/interface.go]: If each plugin that implements `NormalizeScore` should also have a `Score` method, why not embed ScorePlugin into NormalizeScorePlugin like this?

```go
type NormalizeScorePlugin interface {
	ScorePlugin

	NormalizeScore(pc *PluginContext, scores NodeScoreList) *Status
}
```

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (draveness) [reply pkg/scheduler/framework/v1alpha1/framework.go]: > That would helps save on memory by not passing a large map around

This is not a problem since passing a map to another method only copies the pointer to the map which is only 8 bytes.

Comment (ahg-g) [reply pkg/scheduler/framework/v1alpha1/framework.go]: > > That would helps save on memory by not passing a large map around
> 
> This is not a problem since passing a map to another method only copies the pointer to the map which is only 8 bytes.

I think the suggestion was to tradeoff parallelism for space. Basically run the plugins sequentially, parallelising across nodes only. If we do this, then we just need to maintain two N-sized arrays instead kN arrays (where k is the number of plugins)

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/interface.go]: Sounds like a good idea.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: SGTM.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: Added.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: 1. Added verification logic in framework initialization to make sure an enabled NormalizeScore plugin is also enabled as a Score plugin.
2. Embed Score interface in NormalizeScore interface to make sure a NormalizeScore plugin is always a Score plugin.

Do you think there is anything else missing here?

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/framework.go]: nit:

```go
		enabledScorePlugins := make(map[string]struct{}, len(f.scorePlugins))
		for _, sp := range f.scorePlugins {
			enabledScorePlugins[sp.Name()] = struct{}{}
```

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/framework.go]:
```
_, exists := enabledScorePlugins[p.Name()]
```

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/framework.go]: nit: s/NormalizeScore/normalize score/
nit: remove the `.` in the end

errors should start with lower case letter

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/framework.go]: nit: s/NormalizeScore/normalize score/
nit: remove the `.` in the end

errors should start with lower case letter

**Review by draveness — COMMENTED** [review]
Is it valid to have different plugins with the same name? Score and NormalizeScore plugin are different structs but use the same name, if it is necessary to allow this case, we can't embed `Score` interface into `NormalizeScore`.

```go
type TestScorePlugin struct {
}

func (*TestScorePlugin) Name() string {
   return "score-plugin"
}

func (*TestScorePlugin) Score(...) {
}

type TestNormalizePlugin struct {
}

func (*TestNormalizePlugin) Name() string {
   return "score-plugin"
}

func (*TestNormalizePlugin) NormalizeScore(...) {
}
```

**Comment by congliu01** [comment]
> Is it valid to have different plugins with the same name? Score and NormalizeScore plugin are different structs but use the same name, if it is necessary to allow this case, we can't embed `Score` interface into `NormalizeScore`.
> 
> ```go
> type TestScorePlugin struct {
> }
> 
> func (*TestScorePlugin) Name() string {
>    return "score-plugin"
> }
> 
> func (*TestScorePlugin) Score(...) {
> }
> 
> type TestNormalizePlugin struct {
> }
> 
> func (*TestNormalizePlugin) Name() string {
>    return "score-plugin"
> }
> 
> func (*TestNormalizePlugin) NormalizeScore(...) {
> }
> ```

No, a name can only be used to refer to a single plugin. Below is the Plugin Registry [code](https://github.com/kubernetes/kubernetes/blob/54afaf23267b6e0bd24503c712bc21c2b96c7042/pkg/scheduler/framework/v1alpha1/registry.go#L35).
```
// Register adds a new plugin to the registry. If a plugin with the same name
// exists, it returns an error.
func (r Registry) Register(name string, factory PluginFactory) error {
	if _, ok := r[name]; ok {
		return fmt.Errorf("a plugin named %v already exists", name)
	}
	r[name] = factory
	return nil
}
```

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/interface.go]: Since now they are merged, how about we name this interface ScoreWithNormalizePlugin since it is not just normalize.

**Comment by congliu01** [comment]
One of the integ test failures (`TestWebhookV1beta1/settings.k8s.io.v1alpha1.podpresets/delete`) is related to a flaky test reported [here](https://github.com/kubernetes/kubernetes/issues/80258).

The other one (`TestVolumeProvision`) is related to [this closed issue](https://github.com/kubernetes/kubernetes/issues/70180). I filed a new issue: https://github.com/kubernetes/kubernetes/issues/80528

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/interface.go]: This is a good idea.
I would also like to discuss another idea: can we just have a single `ScorePlugin` interface which has both `Score` and `NormalizeScore` methods. If `NormalizeScore` is not needed, it can be implemented as a noop. I think this simplifies the documentation as well as the implementation.

Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/interface.go]: We can still keep both Score and NormalizeScore interfaces. I think the real problem here is that NormalizeScore is really part of Score plugin. However, in the [plugin config](https://github.com/kubernetes/kubernetes/blob/2cd5fc54a10623c30b257f904244d560f1fad9a0/pkg/scheduler/apis/config/types.go#L165-L169), we have both Score and NormalizeScore PluginSets. 

My suggestion is to have only Score PluginSet in the config. In the framework init phase, we check if the enabled Score plugin also implements NormalizeScore, if so, we add it to framework's [normalizeScorePlugins](https://github.com/kubernetes/kubernetes/blob/26df547c414d87861e197855c65ee6fb7493d93a/pkg/scheduler/framework/v1alpha1/framework.go#L45).

I prefer to keep this PR as is but I will have a followup PR to implement my suggestion if it sounds good to everyone.

Comment (ahg-g) [reply pkg/scheduler/framework/v1alpha1/interface.go]: I think removing the explicit [NormalizeScore list](https://github.com/kubernetes/kubernetes/blob/2cd5fc54a10623c30b257f904244d560f1fad9a0/pkg/scheduler/apis/config/types.go#L169) from the config makes sense if we agree that we don't need to allow disabling NormalizeScore for a plugin that implements Score and NormalizeScore.

Doing it in a separate PR is fine, though I prefer to do it in this one if there is no objection to the assumption stated above, @bsalamat WDYT?

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (hex108) [inline pkg/scheduler/framework/v1alpha1/framework.go]: We do not need L375~378 since we make sure it exists. We could add the comment to L371. The code will be more clear.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (hex108) [inline pkg/scheduler/framework/v1alpha1/framework.go]: Same as [comment](https://github.com/kubernetes/kubernetes/pull/80383/files#r307218003), we could skip the check. Then there is no error in this func, and we do not need `errCh` to deal with errors.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (ahg-g) [reply pkg/scheduler/framework/v1alpha1/framework.go]: I suggested we add this check in an earlier round of comments, I agree that in practice this may not happen, however since this is a public function and this map is passed down to the function, in theory there could be a bug in the code path leading to here that results in the score list of a plugin not being added or somehow removed from the map. I would rather the system throw an error message than crash in this case.

One thing we can do to avoid those extra checks and simplify the code in generic_scheduler is to make both RunNormalizeScorePlugins and ApplyWeights private functions of the framework and call them directly at the end of RunScorePlugins.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (ahg-g) [reply pkg/scheduler/framework/v1alpha1/framework.go]: ditto comment above.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (congliu01) [reply pkg/scheduler/framework/v1alpha1/framework.go]: I like the idea of making both RunNormalizeScorePlugins and ApplyWeights private functions. However, I'd like to do it in a follow up PR where I will add unit tests for RunScorePlugins.

Comment (hex108) [reply pkg/scheduler/framework/v1alpha1/framework.go]: SGTM

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/interface.go]: Since we have tight coupling between score and normalize, I think we should treat this as a special type of Score plugin.

I suggest we remove the explicit NormalizeScore list from the [plugins struct](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/types.go). A ScoreWithNormalizePlugin can be added in the Score PluginSet.

For each entry in the Score PluginSet, the framework initialization code will basically first make sure that it implements the Score interface and error out if it does not. Next it checks if it also implements the ScoreWithNormalizePlugin interface, if it does, then we add it to the internal normalize plugins list, if not, then we just do nothing.

**Comment by congliu01** [comment]
I ran into a few failures that seem to due to code generation. I have run `./hack/update-codegen`. 

I need help to fix those failures!

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (bsalamat) [reply pkg/scheduler/framework/v1alpha1/interface.go]: This approach is fine with me. Compared to the current approach, this has the limitation of not being able to disable "Normalize Score" plugins independently of Score plugins, but I doubt if users really need to disable "Normalize" only and leave "Score" in place.

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (bsalamat) [reply pkg/scheduler/framework/v1alpha1/interface.go]: @ahg-g +1

**Comment by hex108** [comment]
> I ran into a few failures that seem to due to code generation. I have run ./hack/update-codegen.

Could try `hack/update-all.sh` if it is related to code generation. @congliu01

**Comment by ahg-g** [comment]
@bsalamat  does this really need an API review?

@congliu01 I am sorry, we will need to revert the change that removes the NormalizeScore list from the config and do it in a separate PR since it seems to require an API review, which may take some time...

**Comment by congliu01** [comment]
> @bsalamat does this really need an API review?
> 
> @congliu01 I am sorry, we will need to revert the change that removes the NormalizeScore list from the config and do it in a separate PR since it seems to require an API review, which may take some time...

@ahg-g, 
Right... Let me try to get a clarification from Bobby. If it does require API review, I will just revert.

**Comment by congliu01** [comment]
> > @bsalamat does this really need an API review?
> > @congliu01 I am sorry, we will need to revert the change that removes the NormalizeScore list from the config and do it in a separate PR since it seems to require an API review, which may take some time...
> 
> @ahg-g,
> Right... Let me try to get a clarification from Bobby. If it does require API review, I will just revert.

Confirmed with @bsalamat that the config API change requires API review. So I will go ahead to revert that and defer it to next PR.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/framework.go]: The idea was to defer the removing of the [NormalizeScore pluginset](https://github.com/kubernetes/kubernetes/blob/e298fc723f2597697b770f88ce47fe93507f7463/pkg/scheduler/apis/config/types.go#L169) from the types, which triggered the API review to a separate PR, leave the logic here the same as we agreed on.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (liu-cong) [reply pkg/scheduler/framework/v1alpha1/framework.go]: This was in the API change commit which was reverted.

**Review by draveness — CHANGES_REQUESTED** [review]
Please squash the commits and revert the part causes API review.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (liu-cong) [reply pkg/scheduler/framework/v1alpha1/framework.go]: Resolving since we now only rely on the ScorePlugin plugin set. NormalizeScorePlugin plugin set is going to be removed in next PR.

File: pkg/scheduler/framework/v1alpha1/interface.go
Comment (liu-cong) [reply pkg/scheduler/framework/v1alpha1/interface.go]: NormalizeScore list is still in the API but not used in the code. I am going to remove it from the API in a separate PR. Resolving this discussion.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (liu-cong) [reply pkg/scheduler/framework/v1alpha1/framework.go]: Done.

**Comment by ahg-g** [comment]
/lgtm

Thanks Cong!

File: test/integration/scheduler/framework_test.go
Comment (bsalamat) [inline test/integration/scheduler/framework_test.go]: The test can be significantly improved if we have multiple nodes in the test and the score plugin looks at the node name or label, for example, and gives higher score to one of them. Then, the test checks that the pod is scheduled on the node which the score plugin gives a higher score to.

**Review by bsalamat — COMMENTED** [review]
Thanks, @liu-cong!

File: test/integration/scheduler/framework_test.go
Comment (liu-cong) [reply test/integration/scheduler/framework_test.go]: This logic is already tested in the TestScorePlugin test case. I made the TestNormalizeScorePlugin case as simple as checking both the Score and NormalizeScore functions are called. I think this is good enough given that NormalizeScore is well tested in unit tests, and an end-to-end "score and schedule" behavior is covered in TestScorePlugins.

WDYT?

Comment (bsalamat) [reply test/integration/scheduler/framework_test.go]: It makes sense. Thanks!

File: test/integration/scheduler/framework_test.go
Comment (bsalamat) [inline test/integration/scheduler/framework_test.go]: Is the part "one of them will be scored much higher than the others" correct?

**Review by bsalamat — COMMENTED** [review]
Thanks, @liu-cong! Looks good. Just a minor comment.

File: test/integration/scheduler/framework_test.go
Comment (liu-cong) [reply test/integration/scheduler/framework_test.go]: It's only relevant for the Score plugin. The Score plugin implementation should tell it anyways so I removed the comment.

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Thanks, @liu-cong!

**Review by draveness — APPROVED** [review]
/retest
/lgtm

**Pull request merged**
